src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1078 @@
+#ifndef CEREAL_ARCHIVES_BSON_HPP_
+#define CEREAL_ARCHIVES_BSON_HPP_
+
+#include <stack>
+#include <string>
+#include <vector>
+
+#include <cereal/cereal.hpp>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/types.hpp>
+#include <bsoncxx/types/value.hpp>
+
+namespace cereal {
+/**
+ * A templated stuct containing a bool value that specifies whether the
+ * provided template parameter is a BSON type.
+ */
+template <class BsonT>
+struct is_bson {
+    static constexpr auto value = std::is_same<BsonT, bsoncxx::types::b_double>::value ||
```

Comment:
Many of these types don't have a storage (b_utf8 for instance just holds a string_view).

Not sure how that would work (those fields would only be valid if the underlying buffer's bytes hadn't been moved around).

If we want to support having those types in your struct we should probably return a buffer along with them (with the bytes copied out of the stream)